opendistro-for-elasticsearch/sql
# Support ORDER BY in new SQL engine

*Issue #, if available:*

1. https://github.com/opendistro-for-elasticsearch/sql/issues/123: Order by aggregate function even if it's not in SELECT
2. https://github.com/opendistro-for-elasticsearch/sql/issues/674: Order by inner field of group by
3. https://github.com/opendistro-for-elasticsearch/sql/issues/277: Order by aggregate function
4. https://github.com/opendistro-for-elasticsearch/sql/issues/250: similar as above
5. https://github.com/opendistro-for-elasticsearch/sql/issues/423: similar as above
6. https://github.com/opendistro-for-elasticsearch/sql/issues/402: similar as above

*Description of changes:*

1. Relax `QualifiedName` to `UnresolvedExpression` in `Field` so we can support order by any expression for SQL.
2. Add `AstSortBuilder` that shares `replaceIfAliasOrOrdinal()` logic with `AstAggregationBuilder`.
3. Improve test framework to support comparison test for `ORDER BY` queries.

With these changes, we can support **ORDER BY field, scalar function, aggregate function, alias and ordinal**. Please see the documentation below for more details.

*Documentation*: https://github.com/dai-chen/sql/blob/support-order-by-in-new-sql-engine/docs/user/dql/basics.rst#example-3-ordering-by-aggregate-functions

*Testing*:

1. Add new UT and comparison test for basic cases
2. Add comparison tests for issues above for regression test
3. Bypass legacy `OrderIT` and `dateFunctionNameCaseInsensitiveTest()`: the former tests explain output and the latter tests `DATE_FORMAT` which has different signature in legacy and new engine.


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
